xarray/tests/test_datatree.py

```diff
@@ -538,6 +539,165 @@ def test_setitem_dataarray_replace_existing_node(self):
+            data_vars={
+                "foo": (["x", "y"], np.random.randn(2, 3)),
+            },
+            coords={
+                "x": ("x", np.array([-1, -2], "int64")),
+                "y": ("y", np.array([0, 1, 2], "int64")),
+                "a": ("x", np.array([4, 5], "int64")),
+                "b": np.int64(-10),
+            },
+        )
+        dt = DataTree(data=ds)
+        dt["child"] = DataTree()
+
+        actual = dt.copy(deep=True)
+        actual.coords["x"] = ("x", ["a", "b"])
+        assert_array_equal(actual["x"], ["a", "b"])
+
+        actual = dt.copy(deep=True)
+        actual.coords["z"] = ("z", ["a", "b"])
+        assert_array_equal(actual["z"], ["a", "b"])
+
+        actual = dt.copy(deep=True)
+        with pytest.raises(ValueError, match=r"conflicting dimension sizes"):
+            actual.coords["x"] = ("x", [-1])
+        assert_identical(actual, dt)  # should not be modified
+
+        actual = dt.copy()
+        del actual.coords["b"]
+        expected = dt.reset_coords("b", drop=True)
+        assert_identical(expected, actual)
+
+        with pytest.raises(KeyError):
+            del dt.coords["not_found"]
+
+        with pytest.raises(KeyError):
+            del dt.coords["foo"]
+
+        actual = dt.copy(deep=True)
+        actual.coords.update({"c": 11})
+        expected = dt.assign_coords({"c": 11})
+        assert_identical(expected, actual)
+
+        # regression test for GH3746
+        del actual.coords["x"]
+        assert "x" not in actual.xindexes
+
+        # test that constructors can also handle the `DataTreeCoordinates` object
+        ds2 = Dataset(coords=dt.coords)
+        assert_identical(ds2.coords, dt.coords)
+        da = DataArray(coords=dt.coords)
+        assert_identical(da.coords, dt.coords)
+
+        # DataTree constructor doesn't accept coords= but should still be able to handle DatasetCoordinates
+        dt2 = DataTree(data=dt.coords)
+        assert_identical(dt2.coords, dt.coords)
+
+    def test_inherited(self):
+        ds = Dataset(
+            data_vars={
+                "foo": (["x", "y"], np.random.randn(2, 3)),
+            },
+            coords={
+                "x": ("x", np.array([-1, -2], "int64")),
+                "y": ("y", np.array([0, 1, 2], "int64")),
+                "a": ("x", np.array([4, 5], "int64")),
+                "b": np.int64(-10),
+            },
+        )
+        dt = DataTree(data=ds)
+        dt["child"] = DataTree()
+        child = dt["child"]
+
+        assert set(child.coords) == {"x", "y", "a", "b"}
+
+        actual = child.copy(deep=True)
+        actual.coords["x"] = ("x", ["a", "b"])
+        assert_array_equal(actual["x"], ["a", "b"])
+
+        actual = child.copy(deep=True)
+        actual.coords.update({"c": 11})
+        expected = child.copy(deep=True)
+        expected.coords["c"] = 11
+        # check we have only altered the child node
+        assert_identical(expected.root, actual.root)
+
```

Comment:
Consder adding a test to verify that trying to delete a coordinate defined on the root raises an exception.

Reply:
good point - checked in https://github.com/pydata/xarray/pull/9451/commits/4d352bd392502a57ec3f00565ea96bf8dd8d3792